### PR TITLE
Variable Dimension Halo

### PR DESCRIPTION
**Pull request by sslattery**
Part of #340. Depends on #349

### CODE REVIEW

**Comment by sslattery** [comment]
@kwitaechong @streeve This one is next for #340

File: cajita/unit_test/tstHalo3d.hpp [hunk 1]
```diff
@@ -73,16 +70,9 @@ void checkGather( const int halo_width, const Array& array )
                          j >= owned_space.max( Dim::J ) + halo_width + pad_j ||
                          k < owned_space.min( Dim::K ) - halo_width ||
                          k >= owned_space.max( Dim::K ) + halo_width + pad_k )
-                    {
-                        if ( host_view( i, j, k, l ) != 0.0 )
-                            pass_gather_test = 0;
-                    }
+                        EXPECT_EQ( host_view( i, j, k, l ), 0.0 );
```
Comment (streeve) [inline cajita/unit_test/tstHalo3d.hpp:73]: This is better, but will take much longer until we fix the failing distributions

File: cajita/unit_test/tstHalo3d.hpp [hunk 2]
```diff
@@ -369,7 +352,7 @@ void scatterReduceTest( const ReduceFunc& reduce )
     std::array<bool, 3> is_dim_periodic = { true, true, true };
     auto global_grid =
         createGlobalGrid( MPI_COMM_WORLD, global_mesh, is_dim_periodic,
-                          Cajita::UniformDimPartitioner() );
+                          Cajita::DimBlockPartitioner<3>() );
```
Comment (streeve) [inline cajita/unit_test/tstHalo3d.hpp:355]: namespace not necessary

**Review by streeve — APPROVED** [review]

File: cajita/src/Cajita_Halo.hpp [hunk 3]
```diff
@@ -458,17 +520,30 @@ class Halo
...
-            "steering", buffer_num_element ) );
+        steering.push_back( Kokkos::View<int**, memory_space>(
+            "steering", buffer_num_element, 3 + NumSpaceDim ) );
 
+        // Build steering vector.
+        buildSteeringVector( spaces, value_byte_sizes, buffer_bytes,
+                             buffer_num_element, steering );
+    }
+
+    // Build steering vector.
```
[the remark this answers is not in the document]
Comment (sslattery) [reply cajita/src/Cajita_Halo.hpp:531]: fixed

File: cajita/unit_test/tstHalo3d.hpp (hunk 2 above)
Comment (sslattery) [reply cajita/unit_test/tstHalo3d.hpp:355]: Fixed

File: cajita/unit_test/tstHalo3d.hpp (hunk 1 above)
Comment (sslattery) [reply cajita/unit_test/tstHalo3d.hpp:73]: OK I think these changes to this part of the test probably overwrote your updates when I did the rebase. I am going to revert back to your test structure for now.

**Pull request merged**